broadinstitute/single_cell_portal_core
# Speed up large expression scatter plots 3-4x via bucket cache data (SCP-4642, SCP-4792)

This greatly speeds up gene expression scatter plots that are >= 0.5M points, by using data cached in buckets.

Previously, this data was retrieved from MongoDB via Rails.  Now, building atop infrastructure described in #1634, this data is pulled as a pre-computed array stored in a flat JSON file in a Google bucket.  So it completely bypasses our traditional backend, by expanding our pre-existing use of buckets as a storage medium.  This will theoretically speed up plots that are 1-1.5M points from [46 s to 15 s](https://mixpanel.com/s/36K58R) (3.1x), and plots that are 0.5-1M points from [19 s to 4.5 s](https://mixpanel.com/s/2B2qli) (4.2x).

To test:
1. Enter development Rails console: 
`bin/rails console --environment development`
2. Find your local DE pilot study: 
`study = Study.find_by(name: 'Human milk - differential expression')`
3. Manually adjust default clustering: 
`study.cluster_groups.find_by(name: 'All Cells UMAP').update(has_image_cache: true)`
4. Get study's bucket ID (prints "fc-65379b91-5ded-4d28-8e51-ada209542117" for me): 
`study.bucket_id`
5. Copy test files from staging DE pilot to your development DE pilot:
```
DEST_BUCKET="<your bucket ID from step 5>"; SRC_BUCKET="fc-37c32921-4c79-4767-819d-66f157f871a9"; gcloud storage cp "gs://${SRC_BUCKET}/_scp_internal/cache/expression_scatter/images/_debug_2022-10-25T164028/A1BG-AS1.webp" "gs://${DEST_BUCKET}/_scp_internal/cache/expression_scatter/images/All_Cells_UMAP/A1BG-AS1.webp"; gcloud storage cp "gs://${SRC_BUCKET}/_scp_internal/cache/expression_scatter/images/_debug_2022-10-25T164028/A1BG.webp" "gs://${DEST_BUCKET}/_scp_internal/cache/expression_scatter/images/All_Cells_UMAP/A1BG.webp"; gcloud storage cp "gs://${SRC_BUCKET}/_scp_internal/cache/expression_scatter/data/All_Cells_UMAP/A1BG.json" "gs://${DEST_BUCKET}/_scp_internal/cache/expression_scatter/data/All_Cells_UMAP/"; gcloud storage cp "gs://${SRC_BUCKET}/_scp_internal/cache/expression_scatter/data/All_Cells_UMAP/A1BG-AS1.json" "gs://${DEST_BUCKET}/_scp_internal/cache/expression_scatter/data/All_Cells_UMAP/"
```
6.  In. your Rails console, get your DE pilot's accession:
`study.accession`
8.  In your web browser, sign in to your local SCP
9.  If you haven't already, enable the "progressive_loading" feature flag on your local DE pilot study (Upper right menu, Feature flags, search by your DE pilot's accession, set "progressive loading" to true)
10.  Search for your local DE pilot study ("Human milk - differential expression"), go to its Explore tab
11. Open Network panel
12. Search for gene "A1BG"
13. Confirm the gene expression scatter plot renders
14. In Network panel, filter requests to "storage.googleapis.com method:GET"
15. Confirm there are 2 matching successful GET requests

This satisfies SCP-4642 and SCP-4792.
